Sources/ProjectDescription/Config.swift

```diff
@@ -52,6 +52,10 @@ public struct Config: Codable, Equatable {
         /// Allows to suppress warnings in Xcode about updates to recommended settings added in or below the specified Xcode version. The warnings appear when Xcode version has been upgraded.
         /// It is recommended to set the version option to Xcode's version that is used for development of a project, for example `.lastUpgradeCheck(Version(13, 0, 0))` for Xcode 13.0.0.
         case lastXcodeUpgradeCheck(Version)
+
+        /// Allows setting a custom directory to be used when resolving package dependencies
+        /// This path is passed to `xcodebuild` via the `-clonedSourcePackagesDirPath` argument
+        case clonedSourcePackagesDirPath(Path)
```

Comment:
Out of curiosity, is the use case reusing that directory across CI environments to speed up builds?

Reply:
yeah exactly, I don't expect huge speed improvements here (since the content still needs to be downloaded) but can be optimised when the same set of sources can be shared across a pipeline

Reply:
What do you think about making that conventional? Where are they currently being downloaded? The main motivation for making it conventional is to minimize the configuration surface as much as possible. The more options we have, the more combinations between are possible, and the harder it becomes for users and us to reason about all of them.

Reply:
> Where are they currently being downloaded?

The default from my testing seems to be in the derived data directory but I need to be able to assign a path for caching on CI and its difficult to be sure where that path will be (as well as being outside the build dir - see below)

> What do you think about making that conventional? 

That would be fine, tbh I don't care at all about the location 😄 What I'm interested in is having a standard place that I can cache by providing a fixed path. The disclaimer here is that the CI I'm using (gitlab) uses cache paths relative to the source directory so caching outside of that (e.g. something like `~/.tuist`) would require some extra work (copying files from those directories to somewhere local which can be cached)

Reply:
Re: convention - what about saving it to `Tuist/Cache/SPM`? You could safely `.gitignore` it and we could reuse the `Tuist/Cache` directory if we need to cache something in the project directory.

Reply:
Because cloned repositories of SPM packages can be shared across projects, I think it makes sense for the directory to live under `~/.tuist/Cache`, maybe `~/.tuist/Cache/ClonedPackages`. Then we can allow [customizing](https://github.com/tuist/tuist/blob/main/Sources/TuistSupport/Utils/Environment.swift#L45) that through environment variables for CI environments. We can name the variable `TUIST_CONFIG_CACHE_DIRECTORY`.

Reply:
> I think it makes sense for the directory to live under `~/.tuist/Cache`, maybe `~/.tuist/Cache/ClonedPackages`. 

That sounds like a good plan to me 👍

> Then we can allow [customizing](https://github.com/tuist/tuist/blob/main/Sources/TuistSupport/Utils/Environment.swift#L45) that through environment variables for CI environments. We can name the variable `TUIST_CONFIG_CACHE_DIRECTORY`.

Also really useful. Do you think it's worth allowing changing the full `~/.tuist/` directory via env var?

Reply:
Actually re the 2nd point I see I can do that in config already

Reply:
I agree we should put them in `~/.tuist/Cache/Something` and I like convention over configuration, but for how SPM works I expect you can't use the same folder across projects (but maybe I'm wrong) 🤔

Reply:
thats a good point and from my testing you are right, would something like `~/.tuist/Cache/ClonedPackages/<Workspace Name>` work?

Reply:
> how SPM works I expect you can't use the same folder across projects (but maybe I'm wrong)
That's something I'd validate first. If SPM can reuse caches, then we don't need the `<Workspace Name>` component to it.

Reply:
If that's the case, I'm not sure `Workspace Name` would be enough, as I'm not sure what would happen if you clone the same repo in 2 different folders. Maybe the best option is to cache it locally? 🤔

Reply:
From my testing using a workspace path component is enough to keep the cache valid for each project _but only_ if that workspace is unique for your workspaces. Is there any other unique ID we could use, e.g. a hash of the full path? Something similar to derived data where the folder is `<Target>-<identifier>` e.g. `Workspace-gxpmepmdjonjmlcnxapjyhuefvtd`

Reply:
I'm not sure I'm following, if the argument is `-clonedSourcePackagesDirPath` I'd assume the directory that we pass there will have cloned repositories that are identified by a sha. That's a very standard structure easily reusable across workspaces so I don't see the need for adding any disambiguation component to the path. It should be sufficient with `~/.tuist/Cache/ClonedPackages`.

Reply:
From my tests this directory is reset at each use, so as a test I created `package_a` with a dependency on `swift-composable-architecture` and `package_b` with a dependency on `Alamofire`. I then ran this on each package

```
xcodebuild -resolvePackageDependencies  -clonedSourcePackagesDirPath common_directory
```

The result was:
- After running on A: `common_directory` contained TCA + dependencies in `/repositories` and `/checkouts`
- After running on B: `common_directory` contained _only_ `Alamofire` in `/repositories` and `/checkouts`
- Running A again resets to step 1 etc

Reply:
@wattson12 what are the reasons why you are not using Dependencies.swift instead? with that, this behaviour would be available out of the box

Reply:
> @wattson12 what are the reasons why you are not using Dependencies.swift instead? with that, this behavior would be available out of the box

We were blocked on using `Dependencies.swift` for some packages by issues with tuist's integration, specifically https://github.com/tuist/tuist/issues/3526 and https://github.com/tuist/tuist/issues/3597

I saw this as an intermediate step on our migration to using only `Dependencies.swift`. Now that https://github.com/tuist/tuist/issues/3526 has been fixed it is probably no longer required so this could be closed on next release (I _think_ I can workaround the issue in https://github.com/tuist/tuist/issues/3597 with a combination of `productTypes` and `targetSettings` configuration as well, but haven't had a chance to test that from the latest `main`)